Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+    , currentFileNumber_(0)
+    , currentSubRunNumber_(-1U)
+    , currentEventNumber_(0)
+    , maxEvents_(conf().maxEvents())
+    , verbosityLevel_(conf().verbosityLevel())
+    , binaryFileVersion_(0)
+  {
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "HPGe");
+    rh.reconstitutes<mu2e::STMWaveformDigiCollection,art::InEvent>(myModuleLabel_, "LaBr");
+    rh.reconstitutes<mu2e::STMTestBeamEventInfo,art::InEvent>(myModuleLabel_);
+
+    currentSubRunNumber_ = 0;
+    currentEventNumber_ = 0;
+
+    printAtEvent = 0;
+
+  }
+
+
+  //----------------------------------------------------------------
+  void STMTestBeamDataDetail::readFile(const std::string& filename, art::FileBlock*& fb) {
+
+    // Open the input file
+    currentFileName_ = filename;
+    currentFile_ = new std::ifstream(currentFileName_.c_str(), std::ios::in | std::ios::binary);
+    if (!currentFile_->is_open()) {
+      throw cet::exception("FromSTMTestBeamData") << "A problem opening binary file " << currentFileName_ << std::endl;
+    }
+
+    // Extract the run number and subrun number from the file name
+    std::string delimeter = ".";
+    size_t currentPos = 0;
+    size_t previousPos = currentPos;
+    std::vector<std::string> tokens;
+    while ( (currentPos = currentFileName_.find(delimeter, previousPos)) != std::string::npos) {
+      tokens.push_back(currentFileName_.substr(previousPos, currentPos-previousPos));
+      previousPos = currentPos+1;
+    }
+    tokens.push_back(currentFileName_.substr(previousPos, currentPos-previousPos)); // add the final token which gets missed in the above loop
+    unsigned int n_expected_fields = 6;
+    if (tokens.size() != n_expected_fields) {
+      throw cet::exception("FromSTMTestBeamData") << "Number of fields in filename (" << tokens.size() << ") is not the same number we expected (" << n_expected_fields << "). Filename = " << currentFileName_ << std::endl;
+    }
+
+    // Get the channel from the configuration field
+    std::string configuration = tokens.at(3);
+    if (configuration.find("HPGe") != std::string::npos) {
+      channel_ = STMChannel::HPGe;
+    }
+    else if (configuration.find("LaBr") != std::string::npos) {
+      channel_ = STMChannel::LaBr;
+    }
+    else {
+      throw cet::exception("FromSTMTestBeamData") << "Cannot determine the channel from the configuration field (" << configuration << "). This should contain either \"HPGe\" or \"LaBr\"" << std::endl;
+    }
+
+    // Get the run and subrun numbers from the sequencer
+    std::string sequencer = tokens.at(4);
+    std::string runNo = sequencer.substr(0, 6); // sequencer always has 6 characters for run
+    runNumber_ = std::stoi(runNo);
+
+    if(!art::RunID(runNumber_).isValid()) {
+      throw cet::exception("BADCONFIG", " FromSTMTestBeamData: ")
+        << " fhicl::ParameterSet specifies an invalid runNumber = "<<runNumber_<<"\n";
+    }
+
+    if(runNumber_ < 101000 || runNumber_ > 101999) {
```

Comment:
I just added this check so that we keep within our allowed run numbers for the test beam data. In the future, we won't need this source module and all the run numbering will be done with art and the run database